File: abelkhan_type_csharp_component_msgpack/component/center/svr_msg_handle.cs
/*
 * svr_msg_handle
 * 2020/6/2
 * qianqians
 */

namespace abelkhan
{
    public class svr_msg_handle
    {
        private center_module _center_module;
        private svrmanager svrmng;
        private hubmanager hubmng;

        public svr_msg_handle(modulemng modules, svrmanager svrs, hubmanager hubs)
        {
            svrmng = svrs;
            hubmng = hubs;
            _center_module = new center_module(modules);

            _center_module.onreg_server += reg_server;
        }

        private void reg_server(string type, string hub_type, string name, string ip, int port)
        {
            hubmng.for_each_hub((hubproxy _proxy) =>{
                _proxy.distribute_server_address(type, name, ip, (ushort)port);
            });

            if (type == "hub")
            {
                var _hubproxy = hubmng.reg_hub(_center_module.current_ch, type, hub_type, name);

                svrmng.for_each_svr((svrproxy _proxy) =>{
                    _hubproxy.distribute_server_address(_proxy.type, _proxy.name, _proxy.ip, _proxy.port);
                });
            }

            var _svrproxy = svrmng.reg_svr(_center_module.current_ch, type, hub_type, name, ip, (ushort)port);
            _svrproxy.reg_server_sucess();
        }
    }
}

File: abelkhan_type_component/component/center/svr_msg_handle.ts
import abelkhan = require('../../abelkhan_type/ts/abelkhan');
import center_protcol = require('../../protcol/ts/center');
import svr = require('./svrmanager');
import hub = require('./hubmanager');

export class svr_msg_handle{
    private center_module : center_protcol.center_module;
    private svrmng : svr.svrmanager;
    private hubmng : hub.hubmanager;
    constructor(modules:abelkhan.modulemng, svrs:svr.svrmanager, hubs:hub.hubmanager){
        this.center_module = new center_protcol.center_module(modules);
        this.svrmng = svrs;
        this.hubmng = hubs;

        this.center_module.cb_reg_server = this.reg_server.bind(this);
    }

    public reg_server(type:string, name:string, ip:string, port:number){
        this.hubmng.for_each_hub((_proxy:hub.hubproxy)=>{
            _proxy.distribute_server_address(type, name, ip, port);
        });

        if (type == "hub") {
            let _hubproxy = this.hubmng.reg_hub (this.center_module.current_ch);

            this.svrmng.for_each_svr((_proxy:svr.svrproxy) =>{
                _hubproxy.distribute_server_address(_proxy.type, _proxy.name, _proxy.ip, _proxy.port);
            });
        }

        let _svrproxy = this.svrmng.reg_svr(this.center_module.current_ch, type, name, ip, port);
		_svrproxy.reg_server_sucess();
    }
}